core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CloneWorkspaceFileTransferMonitor.scala

```diff
@@ -0,0 +1,78 @@
+
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+
+
+object CloneWorkspaceFileTransferMonitor {
+  def props(dataSource: SlickDataSource, gcsDAO: GoogleServicesDAO, initialDelay: FiniteDuration, pollInterval: FiniteDuration)(implicit executionContext: ExecutionContext, cs: ContextShift[IO]): Props = {
+    Props(new CloneWorkspaceFileTransferMonitorActor(dataSource, gcsDAO, initialDelay, pollInterval))
+  }
+
+  sealed trait CloneWorkspaceFileTransferMonitorMessage
+  case object CheckAll extends CloneWorkspaceFileTransferMonitorMessage
+}
+
+class CloneWorkspaceFileTransferMonitorActor(val dataSource: SlickDataSource, val gcsDAO: GoogleServicesDAO, val initialDelay: FiniteDuration, val pollInterval: FiniteDuration)(implicit executionContext: ExecutionContext, cs: ContextShift[IO]) extends Actor with LazyLogging {
+
+  context.system.scheduler.scheduleWithFixedDelay(initialDelay, pollInterval, self, CheckAll)
+
+  override def receive = {
+    case CheckAll => checkAll()
+  }
+
+  private def checkAll() = {
+    for {
+      pendingTransfers <- dataSource.inTransaction { dataAccess =>
+        dataAccess.cloneWorkspaceFileTransferQuery.listPendingTransfers()
+      }
+      _ <- pendingTransfers.toList.traverse { pendingTransfer =>
+        IO.fromFuture(IO(copyBucketFiles(pendingTransfer)))
+      }.unsafeToFuture()
+    } yield()
+  }
+
+  private def copyBucketFiles(pendingCloneWorkspaceFileTransfer: PendingCloneWorkspaceFileTransfer): Future[List[Option[StorageObject]]] = {
+    for {
+      objectsToCopy <- gcsDAO.listObjectsWithPrefix(pendingCloneWorkspaceFileTransfer.sourceWorkspaceBucketName, pendingCloneWorkspaceFileTransfer.copyFilesWithPrefix, Option(pendingCloneWorkspaceFileTransfer.destWorkspaceGoogleProjectId)).recoverWith {
+        case e: RawlsExceptionWithErrorReport if e.errorReport.statusCode == Option(StatusCodes.Forbidden) => {
+          logger.warn(s"403 received when listing objects in ${pendingCloneWorkspaceFileTransfer.sourceWorkspaceBucketName} before copying to ${pendingCloneWorkspaceFileTransfer.destWorkspaceBucketName}: $e")
+          Future.failed(e)
+        }
+      }
+      copiedObjects <- Future.traverse(objectsToCopy) { objectToCopy =>
+        gcsDAO.copyFile(pendingCloneWorkspaceFileTransfer.sourceWorkspaceBucketName, objectToCopy.getName, pendingCloneWorkspaceFileTransfer.destWorkspaceBucketName, objectToCopy.getName, Option(pendingCloneWorkspaceFileTransfer.destWorkspaceGoogleProjectId)).recoverWith {
+          case e: RawlsExceptionWithErrorReport if e.errorReport.statusCode == Option(StatusCodes.Forbidden) => {
+            logger.warn(s"403 received when copying [${pendingCloneWorkspaceFileTransfer.sourceWorkspaceBucketName}/${objectToCopy}] to [${pendingCloneWorkspaceFileTransfer.destWorkspaceBucketName}]: $e")
+            Future.failed(e)
```

Comment:
Do we need to worry about collisions in the case where some files fail to copy?

Reply:
Good question, from what I can tell in the docs, repeated attempts to copy the same file will simply overwrite the file that was copied the last time so any collisions should be fine.